Mathlib/Tactic/ComputeDegree.lean

```diff
@@ -4,37 +4,461 @@ Released under Apache 2.0 license as described in the file LICENSE.
+`WithBot ℕ` is the tactic's guess of the degree, and `r` is the coefficient's guess of the
+leading coefficient of `f`.
+
+`compute_degree` applies `norm_num` to the left-hand side of all side goals, trying to clos them.
+
+The variant `compute_degree!` first applies `compute_degree`.
+Then it uses `norm_num` on all the whole remaining goals and tries `assumption`.
+-/
+syntax (name := computeDegree) "compute_degree" "!"? "-debug"? : tactic
+
+@[inherit_doc computeDegree]
+macro "compute_degree!" "-debug"? : tactic => `(tactic| compute_degree ! -debug)
+@[inherit_doc computeDegree]
+macro "compute_degree!" : tactic => `(tactic| compute_degree !)
+
+elab_rules : tactic | `(tactic| compute_degree $[!%$bang]? $[-debug%$dbg]?) => focus do
+  let dbg := dbg.isSome
+  let goal := ← getMainGoal
+  let gt := ← goal.getType''
+  let deg := match gt.eq? with
+    | some (_, _, rhs) => some rhs
+    | _ => none
+  let twoH := twoHeadsArgs gt
+  match twoH with
+    | (.anonymous, na, _, _) => throwError
+        (m!"'compute_degree' inapplicable.  The goal\n\n   {gt}\n\nis expected to be '≤' or " ++
+          m!"'=', instead of '{na}'.")
+    | (na, .anonymous, _, _) => throwError
+        (m!"'compute_degree' inapplicable.  The LHS of\n\n   {gt}\n\nbegins with '{na}'.  " ++
+          m!"There is support only for\n\n   'natDegree'   'degree'   or   'coeff'")
+    | _ =>
+      let lem := dispatchLemma twoH
+      if dbg then logInfo f!"'compute_degree' first applies lemma '{lem.getString}'"
+      let mut (gls, static) := (← goal.applyConst lem, [])
+      while (! gls == []) do (gls, static) := ← splitApply gls static
+      let rfled := ← try_rfl static
+      setGoals rfled
+      --  simplify the left-hand sides, since this is where the degree computations leave
+      --  expressions such as `max (0 * 1) (max (1 + 0 + 3 * 4) (7 * 0))`
+      evalTactic (← `(tactic| try any_goals conv_lhs => (simp only [Nat.cast_withBot]; norm_num)))
+      -- we save in `final_gls` the list of goals *before* `norm_num` has a chance
+      -- to reduce them to `False`, so that we can get an informed error message
+      let final_gls := ← getGoals
+      if bang.isSome then evalTactic (← `(tactic| try (any_goals norm_num <;> try assumption)))
+      done <|> unless deg.isNone do
+        -- we extract, from the list `final_gls` of stored goals, the ones that are now `False`
+        let false_goals := ← final_gls.filterM fun mv_before => do
+          let mv_after? := (← getMCtx).userNames.find? (← mv_before.getTag)
+          match mv_after? with
+            | some mv_after => return ((← mv_after.getType'') == (Expr.const ``False []))
+            | none => return false
+        -- we compute the error message
+        let errors := miscomputedDegree? deg.get! (← false_goals.mapM (MVarId.getType'' ·))
+        if errors.length != 0 then
+          throwError (Lean.MessageData.joinSep
+            (m!"The given degree is '{deg}'.  However,\n" :: errors) "\n")
```

Comment:
1. This removes the debug flag and switches to using `trace[Tactic.compute_degree]`
2. This switches to `let goal ← getMainGoal` notation in a few places instead of `let goal := ← getMainGoal`
3. The two error cases in the `match` seemed backwards. I also removed mentioning the head symbol since a user can see it from the goal state easily enough (and it's wrong to mention `.anonymous` if there's no head symbol). Note the `{""}` trick for continuing an `m!` string.
4. The `while` can use `!=`
5. There was a missing `withMainContext` around the whole `do` expression, which is needed for `getType''`
6. The tag-based goal matching when assembling the miscomputed degree error message wasn't robust, so I changed it to something that is guaranteed to work.
```suggestion
elab_rules : tactic | `(tactic| compute_degree $[!%$bang]?) => focus <| withMainContext do
  let goal ← getMainGoal
  let gt ← goal.getType''
  let deg? := match gt.eq? with
    | some (_, _, rhs) => some rhs
    | _ => none
  let twoH := twoHeadsArgs gt
  match twoH with
    | (_, .anonymous, _) => throwError
        (m!"'compute_degree' inapplicable. The goal{indentD gt}\nis expected to be '≤' or '='.")
    | (.anonymous, _, _) => throwError
        (m!"'compute_degree' inapplicable. The LHS must be an application of {
          ""}'natDegree', 'degree', or 'coeff'.")
    | _ =>
      let lem := dispatchLemma twoH
      trace[Tactic.compute_degree] f!"'compute_degree' first applies lemma '{lem.getString}'"
      let mut (gls, static) := (← goal.applyConst lem, [])
      while gls != [] do (gls, static) ← splitApply gls static
      let rfled ← try_rfl static
      setGoals rfled
      --  simplify the left-hand sides, since this is where the degree computations leave
      --  expressions such as `max (0 * 1) (max (1 + 0 + 3 * 4) (7 * 0))`
      evalTactic
        (← `(tactic| try any_goals conv_lhs => (simp only [Nat.cast_withBot]; norm_num)))
      if bang.isSome then
        let mut false_goals : Array MVarId := #[]
        let mut new_goals : Array MVarId := #[]
        for g in ← getGoals do
          let gs' ← run g do evalTactic (← `(tactic| try (any_goals norm_num <;> try assumption)))
          new_goals := new_goals ++ gs'.toArray
          if ← gs'.anyM fun g' => g'.withContext do return (← g'.getType'').isConstOf ``False then
            false_goals := false_goals.push g
        setGoals new_goals.toList
        if let some deg := deg? then
          let errors := miscomputedDegree? deg (← false_goals.mapM (MVarId.getType'' ·)).toList
          unless errors.isEmpty do
            throwError Lean.MessageData.joinSep
              (m!"The given degree is '{deg}'.  However,\n" :: errors) "\n"
```